server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,262 @@
+            # we also strongly incentivise the algorithm to give both teams the same number of players
+            containers.append(Container(5000 * len(s.players) + s.cumulated_rating, [s]))
+
+        containers.sort(key=lambda c: c.rating)
+        elem1 = containers.pop()
+        elem2 = containers.pop()
+        while True:
+            #  elem1 is always bigger than elem2
+            container = Container(elem1.rating - elem2.rating, [elem1, elem2])
+            containers.append(container)
+            containers.sort(key=lambda c: c.rating)
+            elem1 = containers.pop()
+            try:
+                elem2 = containers.pop()
+            except IndexError:
+                break
+        self._logger.debug("Rating disparity: %s", elem1.rating)
+
+        team_a = []
+        team_b = []
+        containers_a = []
+        containers_b = []
+        containers_a.append(elem1.content[0])
+        containers_b.append(elem1.content[1])
+        while len(containers_a) > 0 or len(containers_b) > 0:
+            for e in containers_a:
+                if len(e.content) == 2:
+                    containers_a.append(e.content[0])
+                    containers_b.append(e.content[1])
+                else:
+                    team_a.append(e.content[0])
+                containers_a.remove(e)
+            for e in containers_b:
+                if len(e.content) == 2:
+                    containers_b.append(e.content[0])
+                    containers_a.append(e.content[1])
+                else:
+                    team_b.append(e.content[0])
+                containers_b.remove(e)
+        return team_a, team_b
+
+    def _searches_by_size(self, searches: List[Search]) -> Dict[int, List[Search]]:
+        searches_by_size: Dict[int, List[Search]] = defaultdict(list)
+
+        for search in searches:
+            size = len(search.players)
+            searches_by_size[size].append(search)
+
+        self._logger.debug("participating searches by player size:")
+        for i in range(self.team_size, 0, -1):
+            self._logger.debug("%i players: %s", i, searches_by_size[i])
+        return searches_by_size
+
+    def _find_most_balanced_filler(self, avg, search, searches_dict):
+        """
+        If we simply fetch the highest/lowest rated single player search we may overshoot our
+        goal to get the most balanced teams, so we try them all until we find the one that brings
+        us closest to the rating average i.e. balanced teams
+        If there is no single player search we have hit a search combination that is impossible to
+        separate into two teams e.g. (3, 3, 2) for 4v4
+        """
+        team_avg = search.average_rating
+        if not searches_dict[1]:
+            self._logger.warning("given searches are impossible to split in even teams because of party sizes")
+            raise AssertionError
+        searches_dict[1].sort(key=lambda s: s.cumulated_rating, reverse=True)
```

Comment:
Maybe put in the docstring that this function modifies its arguments

Reply:
TBH I'm not sure why this function is so complicated. Can't you just iterate through this list and find the best match according to some metric? I feel like this should look very similar to a "max" implementation:
```python3
if not the_list:
    return None

iterator = iter(the_list)
best = next(iterator)
for item in iterator:
    if item > best:
        best = item
return best
```

Reply:
It's partly because the code looks messier because it's not using an inteator. The other part of the answer is that the condition to abort is not so easy. I can't check if a metric is fulfilled, but I need to check for better solutions until I find a worse one and know because of this what the best solution was.

Reply:
Why do you need to find a worse one? You only care about the best right?

Reply:
I need to find the player that gives me the team that is closest to the intended rating. It's basically an optimization problem. The perfect solution (getting exactly the intended average) will generally not happen. So I have to try them one by one as long as the solution gets incrementally better. I only know what the best solution is, when I can be sure that all following candidates will provide a worse solution.